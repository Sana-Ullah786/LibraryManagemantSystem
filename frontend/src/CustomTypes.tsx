/**
 * This file contains custom data types and interfaces that are generic enough to be used multiple times throughout the project
 */

export interface AuthorDetails {
  first_name?: string | null;
  last_name?: string | null;
  date_of_birth?: string | null;
  date_of_death?: string | null;
}

export interface Author extends AuthorDetails {
  id: number;
}

export type UserLoginData = {
  username: string;
  password: string;
};

export type Tokens = {
  access_token: string;
  refresh_token: string;
};

// The attributes of a book
export interface BookBase {
  title: string;
  description: string;
  isbn: string;
  dateOfPublication: Date;
}

export interface BookOut extends BookBase {
  authorIds: number[];
  genreIds: number[];
  languageId: number;
  numberOfCopies: number;
}

export interface BookSaved extends BookBase {
  id: number;
  authors: Author[];
  language: Language;
  genres: Genre[];
}

//Genre Interfaces


export interface Genre {
  id: number;
  genre: string ;
}

export interface GenreDetails {
  id: number;
  genre: string;
}

export interface ErrorObject {
  status: number;
  message: string;
}

export interface GenreMap {
  [id: number]: Genre;
}

export interface Language {
  id: number;
  language: string;
}
//Signup Attrubutes
export interface SignupData {
  email: string;
  username: string;
  password: string;
  first_name: string;
  last_name: string;
  contact_number: string;
  address: string;
}

//Language Attributes

export interface LanguageDetails {
  language: string | null;
  // Add other properties specific to the LanguageDetails type
}
export interface Language {
  id: number;
  language: string;
}
