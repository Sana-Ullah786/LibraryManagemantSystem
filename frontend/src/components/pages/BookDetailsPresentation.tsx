import { Book } from '../../CustomTypes'
import { Link } from 'react-router-dom'
import { Author } from '../../CustomTypes';
import LibrarianLinks from '../LibrarianLinks';
import { GenreMap, Language } from '../../CustomTypes';

type BookDetailsPresentationProps = {
    isLibrarian: boolean;
    showLinks?: boolean;
    id: number;
    url: string;
    book: Book;
    author?: Author;
    language?: Language;
    genres: GenreMap;
}

const NOT_FOUND = "Not Found";

export const BookDetailsPresentation = (props: BookDetailsPresentationProps) => {

    /*
    * BookDetailsPresentation is responsible for displaying all the details of book.
    * It receives the data as properties and does not have to perform any api calls or calculations.
    */


    return (

        <div>

            <h1>
                <p style={{ display:"inline" }} data-testid="titleHeading"> Title: </p>
                <p style={{ display:"inline" }} data-testid="title"> {props.book.title} </p>    
            </h1>
            {
                props.isLibrarian && props.showLinks && (
                    <>
                        <LibrarianLinks url={props.url} />
                        <br />
                    </>
                )
            }
            <p style={{ display:"inline" }} data-testid="authorHeading">
                <strong>Author: </strong>
            </p>
            <p data-testid="author" style={{ display:"inline" }}>
                {
                    props.author?
                    <Link to={"../authors/" + props.book.authorId}>
                        {props.author.last_name + ", " + props.author.first_name}
                    </Link>
                    : NOT_FOUND
                }
            </p>
            <br></br>
            <br></br>
            <p style={{ display:"inline" }} data-testid="summaryHeading"><strong>Summary: </strong></p>
            <p data-testid="summary" style={{ display:"inline" }}> {props.book.summary}</p>
            <br></br>
            <br></br>
            <p style={{ display:"inline" }} data-testid="isbnHeading"><strong>ISBN: </strong></p>
            <p data-testid="isbn" style={{ display:"inline" }}> {props.book.isbn}</p>
            <br></br>
            <br></br>
            <p style={{ display:"inline" }} data-testid="languageHeading"><strong>Language: </strong></p>
            <p data-testid="language" style={{ display:"inline" }}>
                {
                    props.language !== undefined?
                    props.language.name
                    : NOT_FOUND
                }
            </p>
            <br></br>
            <br></br>
            <p style={{ display:"inline" }} data-testid="genreHeading"><strong>Genre: </strong></p>
            <p data-testid="genres" style={{ display:"inline" }}> {Object.values(props.genres).map((genre) => genre.name).join(", ")}</p>

        </div>
    );


}

// By default, links are always shown if the user is librarian
BookDetailsPresentation.defaultProps = {showLinks: true}

