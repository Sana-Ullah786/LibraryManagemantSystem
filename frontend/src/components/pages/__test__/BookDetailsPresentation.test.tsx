import { render, cleanup } from "@testing-library/react";
import { BookDetailsPresentation } from "../BookDetailsPresentation";
import { Book, AuthorAttributes, Genre, Language } from "../../../CustomTypes";
import { StaticRouter } from "react-router-dom";
import renderer from "react-test-renderer";
//TODO: Changes to the custom types file means that these test cases won't work anymore and will need to be changed
const book: Book = {
  id: 1,
  title: "title",
  authorId: 1,
  description: "description",
  isbn: "1231231231231",
  languageId: 1,
  genreIds: [1, 2],
};

const author: AuthorAttributes = {
  id: 1,
  first_name: "firstname",
  last_name: "lastname",
  birth_date: "dateofbirth",
  death_date: "dateofdeath",
};

const language: Language = {
  id: 1,
  name: "English",
};

const genres: Genre[] = [
  {
    id: 1,
    name: "Fantasy",
  },
  {
    id: 2,
    name: "Drama",
  },
];

// Cleans up the DOM after each test
afterEach(cleanup);

/**
 * These tests check if the books page renders with correct fomat and text for
 * book attribute headings
 */
describe("Tests correct correct rendering of style and headings", () => {
  /**
   * This test checks if the component renders without any errors.
   */
  test("renders without error", () => {
    render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );
  });

  /**
   * This test checks if the text of title heading is correct
   */
  test("renders correct heading for title", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("titleHeading")).toHaveTextContent("Title:");
  });

  /**
   * This test checks if the text of title heading is correct
   */
  test("renders correct heading for author", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("authorHeading")).toHaveTextContent("Author:");
  });

  /**
   * This test checks if the text of description heading is correct
   */
  test("renders correct heading for description", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("descriptionHeading")).toHaveTextContent("Description:");
  });

  /**
   * This test checks if the text of isbn heading is correct
   */
  test("renders correct heading for isbn", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("isbnHeading")).toHaveTextContent("ISBN:");
  });

  /**
   * This test checks if the text of language heading is correct
   */
  test("renders correct heading for language", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("languageHeading")).toHaveTextContent("Language:");
  });

  /**
   * This test checks if the text of genre heading is correct
   */
  test("renders correct heading for genre", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("genreHeading")).toHaveTextContent("Genre:");
  });

  /**
   * This test checks if the child of author heading is strong
   */
  test("renders correct style for author heading", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    const authorHeading = getByTestId("authorHeading");
    expect(authorHeading.firstElementChild?.tagName).toBe("STRONG");
  });

  /**
   * This test checks if the child of genre heading is strong
   */
  test("renders correct style for genre heading", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    const genreHeading = getByTestId("genreHeading");
    expect(genreHeading.firstElementChild?.tagName).toBe("STRONG");
  });
});

/**
 * These tests check if correct data is displayed for each book attribute
 */
describe("Tests correct data displayed for each heading", () => {
  /**
   * This test checks if book title is correctly displayed
   */
  test("renders correct title for book", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("title")).toHaveTextContent(book.title);
  });

  /**
   * This test checks if author text is correct
   */
  test("renders correct author name", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );
    expect(getByTestId("author")).toHaveTextContent(
      author.last_name + ", " + author.first_name
    );
  });

  /**
   * This test checks if description has correct text
   */
  test("renders correct description text", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );
    expect(getByTestId("description")).toHaveTextContent(book.description);
  });

  /**
   * This test checks if language has correct text
   */
  test("renders correct language text", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );
    expect(getByTestId("language")).toHaveTextContent(language.name);
  });

  /**
   * This test checks if genres have correct text
   */
  test("renders correct genres text", () => {
    const { getByTestId } = render(
      <StaticRouter>
        <BookDetailsPresentation
          book={book}
          author={author}
          language={language}
          genres={genres}
        />
      </StaticRouter>
    );

    expect(getByTestId("genres")).toHaveTextContent(
      Object.values(genres)
        .map((genre) => genre.name)
        .join(", ")
    );
  });
});

/**
 * These tests check the snapshot for the rendered elements.
 */
describe("Snapshot tests", () => {
  /**
   * This test makes sure that the snapshot produced is valid and up to date.
   * It detects any changes in user interface and reports them.
   */
  test("produces correct snapshot", () => {
    const tree = renderer
      .create(
        <StaticRouter>
          <BookDetailsPresentation
            book={book}
            author={author}
            language={language}
            genres={genres}
          />
        </StaticRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
