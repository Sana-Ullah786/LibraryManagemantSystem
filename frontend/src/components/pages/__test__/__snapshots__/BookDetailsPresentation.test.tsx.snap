// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests produces correct snapshot 1`] = `
<div>
  <h1>
    <p
      data-testid="titleHeading"
      style={
        Object {
          "display": "inline",
        }
      }
    >
       Title: 
    </p>
    <p
      data-testid="title"
      style={
        Object {
          "display": "inline",
        }
      }
    >
       
      title
       
    </p>
  </h1>
  <p
    data-testid="authorHeading"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <strong>
      Author: 
    </strong>
  </p>
  <p
    data-testid="author"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <a
      href="/authors/1"
      onClick={[Function]}
    >
      lastname, firstname
    </a>
  </p>
  <br />
  <br />
  <p
    data-testid="summaryHeading"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <strong>
      Summary: 
    </strong>
  </p>
  <p
    data-testid="summary"
    style={
      Object {
        "display": "inline",
      }
    }
  >
     
    summary
  </p>
  <br />
  <br />
  <p
    data-testid="isbnHeading"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <strong>
      ISBN: 
    </strong>
  </p>
  <p
    data-testid="isbn"
    style={
      Object {
        "display": "inline",
      }
    }
  >
     
    1231231231231
  </p>
  <br />
  <br />
  <p
    data-testid="languageHeading"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <strong>
      Language: 
    </strong>
  </p>
  <p
    data-testid="language"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    English
  </p>
  <br />
  <br />
  <p
    data-testid="genreHeading"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <strong>
      Genre: 
    </strong>
  </p>
  <p
    data-testid="genres"
    style={
      Object {
        "display": "inline",
      }
    }
  >
     
    Fantasy, Drama
  </p>
</div>
`;
